### PR TITLE
fix(ngcc): libraries using spread in object literals cannot be processed

### LINKED ISSUE #34659: ngcc: libraries using spread operator in object literals cannot be processed
Consider a library that uses a shared constant for host bindings. e.g.

```ts
export const BASE_BINDINGS= {
  '[class.mat-themed]': '_isThemed',
}

----

@Directive({
  host: {...BASE_BINDINGS, '(click)': '...'}
})
export class Dir1 {}

@Directive({
  host: {...BASE_BINDINGS, '(click)': '...'}
})
export class Dir2 {}
```

Previously when these components were shipped as part of the library to NPM, consumers were able to consume `Dir1` and `Dir2`. No errors showed up.

Now with Ivy, when ngcc tries to process the library, an error will be thrown. The error is stating that the host bindings should be an object (which they obviously are). This happens because TypeScript transforms the object spread to individual `Object.assign` calls (for compatibility).

The partial evaluator used by the `@Directive` annotation handler is unable to process this expression because there is no integrated support for `Object.assign`. In View Engine, this was not a problem because the `metadata.json` files from the library were used to compute the host bindings.

### Why should Object.assign work?

Currently ngtsc works with object spread operators. Just ngcc which runs into the transpiled `Object.assign` calls will not support object spreads. This results in inconsistent behavior.

Also, this previously worked in View Engine, so it would help avoiding a breaking change.

### How could it be fixed?

We could add another `BuiltinFn` into the partial evaluator. It will be somewhat difficult to detect whether a expression with `Object.assign` is actually resolving to the default ES6 `Object.assign`.

Since the default libs are disabled in the ngcc program, we cannot simply use the type checker. Joost had the idea of providing these types on our own in the program. That would make the detection easy and also work for ES5 files that rely on the `Object.assign` polyfill.

### PR DESCRIPTION
**Pull request by devversion**
Consider a library that uses a shared constant for host bindings. e.g.

```ts
export const BASE_BINDINGS= {
  '[class.mat-themed]': '_isThemed',
}

----

@Directive({
  host: {...BASE_BINDINGS, '(click)': '...'}
})
export class Dir1 {}

@Directive({
  host: {...BASE_BINDINGS, '(click)': '...'}
})
export class Dir2 {}
```

Previously when these components were shipped as part of the
library to NPM, consumers were able to consume `Dir1` and `Dir2`.
No errors showed up.

Now with Ivy, when ngcc tries to process the library, an error
will be thrown. The error is stating that the host bindings should
be an object (which they obviously are). This happens because
TypeScript transforms the object spread to individual
`Object.assign` calls (for compatibility).

The partial evaluator used by the `@Directive` annotation handler
is unable to process this expression because there is no
integrated support for `Object.assign`. In View Engine, this was
not a problem because the `metadata.json` files from the library
were used to compute the host bindings.

To fix this, we add support for `Object.assign`. Unfortunately there
is no safe and easy way to detect if a property access resolves to
the `assign` method of the global `Object` variable. To ensure that
the partial evaluator can resolve both the "lhs" and "rhs" of the property
access, we add the needed `Object.assign` types to the ngcc program.

Fixes #34659

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts [hunk 1]
```diff
@@ -367,7 +367,13 @@ export class StaticInterpreter {
       return lhs[rhs];
     } else if (lhs instanceof Reference) {
       const ref = lhs.node;
-      if (this.host.isClass(ref)) {
+      const identifier = identifierOfNode(ref);
+      // Handles `Object.assign`. The interpreter intends to support the spread operator, but
+      // should also handle a TypeScript transpiled spread object literal. TypeScript transforms
+      // the spread operator into `Object.assign` calls when targeting anything older than ES2018.
+      if (identifier && identifier.text === 'Object' && rhs === 'assign') {
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts]: This is still unsafe. I'm not sure how we can reliably ensure that this `Reference` is the global `Object`. I don't want to check if the source-file that declares `Reference#node` matches the name of the fake source file we added to the program. This is specific to ngcc, and the isolated interpreter should not know about this implementation detail.

Any ideas?

Comment (devversion) [reply packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts]: Improved by using a global definition that can be identified.

**Review by gkalpak — COMMENTED** [review]
TBH, I would expect (and prefer) `Object.assign()` to be treated as a [BuiltinFn][1] and `__assign()` to be treated as a [TsHelperFn][2]. That would also ensure that fixing/improving the respective features (`BuiltinFn`/`TsHelperFn`) will apply to `assign` too.

(For example, there is work in progress to support imported helpers (e.g. `tslib.__<helper>()`). With the current implementation, this fix should be applied separately to `assign`.

[1]: https://github.com/angular/angular/blob/725148a44/packages/compiler-cli/src/ngtsc/partial_evaluator/src/result.ts#L56
[2]: https://github.com/angular/angular/blob/725148a44/packages/compiler-cli/src/ngtsc/reflection/src/host.ts#L334

**Comment by devversion** [comment]
@gkalpak Thanks for your initial feedback. I initially was trying to use a `BuiltinFn`, but it turned out that it doesn't even get to [`accessHelper`](https://github.com/angular/angular/blob/725148a44db896f919fc0b6d7e256dbc35af3ec6/packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts#L328) which runs the built-in functions. This is because `Object` will resolve to a `DynamicValue`. This might be avoidable with the global type definition, or by introducing a new way that allows mapping identifiers to `ts.Node`'s. This might be useful for your PR that fixes inlined ts helper functions too. I chatted with Alex about this, and I'll be experiment with this.

In general, I agree that it would be good to handle `__assign` through a `TsHelperFn`. Though, for `Object.assign` I feel slightly different. I think we don't want to add support for it in the `Interpreter`, but rather only handle it in ngcc (with the reason of supporting transpiled libraries). If we'd add it to the interpreter, then I could imagine us arguing adding more functions like `Object.keys` etc. 

Also I wanted to mention that I think that most of the evaluation code for built-in functions or helper functions could be simplified, by providing a way to transform the call expression into a corresponding TS expression. e.g. `__spreadArray(a, b)` becomes `[...a, ...b]`. The evaluator will then handle it as expected without us having to worry about the implementation.

**Comment by gkalpak** [comment]
> I initially was trying to use a `BuiltinFn`, but it turned out that it doesn't even get to `accessHelper` [...]
> This might be avoidable with the global type definition [...]

Yes, I believe so.

##
> or by introducing a new way that allows mapping identifiers to `ts.Node`'s. This might be useful for your PR that fixes inlined ts helper functions too.

Indeed. Looking forward to see what you'll come up with.

##
> I think we don't want to add support for `Object.assign` in the `Interpreter`, but rather only handle it in ngcc (with the reason of supporting transpiled libraries).

TBH, I don't see why :grin:
`Object.assign()` is not much different than other `BuiltinFn`s (such as `Array.prototype.slice()`) and if there are `Object.assign()` instances that are _not_ a result of transpilation, it would be weird to support them in ngcc, but not ngtsc.

##
> If we'd add it to the interpreter, then I could imagine us arguing adding more functions like `Object.keys` etc. 

Which is bad, because... 😕 :grin:

File: packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts [hunk 2]
```diff
@@ -8,29 +8,45 @@
...
+ * Resolves the specified builtin declaration to a resolved value. For example,
+ * the known JavaScript global `Object` will resolve to a `Map` that provides the
+ * `assign` method with a builtin function. This enables evaluation of `Object.assign`.
+ */
+export function resolveBuiltinDeclaration(builtinDecl: BuiltinDeclaration): ResolvedValue {
+  switch (builtinDecl) {
+    case BuiltinDeclaration.JsGlobalObject:
+      return new Map([['assign', new ObjectAssignBuiltinFn()]]);
+    default:
+      throw new Error(`Cannot resolve unknown builtin declaration ${builtinDecl}`);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts]:
```suggestion
      throw new Error(`Cannot resolve unknown builtin declaration ${BuiltinDeclaration[builtinDecl]}`);
```

(gives the name, rather than the enum index)

File: packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts [hunk 3]
```diff
@@ -8,29 +8,45 @@
...
+  }
+}
+
 export class ArraySliceBuiltinFn extends BuiltinFn {
-  constructor(private node: ts.Node, private lhs: ResolvedValueArray) { super(); }
+  constructor(private lhs: ResolvedValueArray) { super(); }
 
-  evaluate(args: ResolvedValueArray): ResolvedValue {
+  evaluate(node: ts.Node, args: ResolvedValueArray): ResolvedValue {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts]: This is a `ts.CallExpression`, right? We can type it more narrowly.

File: packages/compiler-cli/src/ngtsc/reflection/src/host.ts [hunk 4]
```diff
@@ -395,6 +409,12 @@ export interface BaseDeclaration<T extends ts.Declaration = ts.Declaration> {
    * TypeScript reference to the declaration itself, if one exists.
    */
   node: T|null;
+
+  /**
+   * If set, describes the type of builtin JavaScript declaration this
+   * declaration resolves to.
+   */
+  builtinDecl?: BuiltinDeclaration
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/reflection/src/host.ts]: In the ngtsc codebase, we don't use optional structure members (at least not often). Additionally, I think `Decl` as a suffix is a bit redundant given this is in a "Declaration" interface.
```suggestion
  builtin: BuiltinDeclaration|null;
```

File: packages/compiler-cli/src/ngtsc/partial_evaluator/test/utils.ts [hunk 5]
```diff
@@ -16,17 +16,20 @@ import {getDeclaration, makeProgram} from '../../testing';
...
-  const {program, options, host} = makeProgram([
-    {name: absoluteFrom('/entry.ts'), contents: `${code}; const target$ = ${expr};`},
-    ...supportingFiles
-  ]);
+  const {program, options, host} = makeProgram(
+      [
+        {name: absoluteFrom('/entry.ts'), contents: `${code}; const target$ = ${expr};`},
+        ...supportingFiles
+      ],
+      undefined, undefined, checkForErrors);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/partial_evaluator/test/utils.ts]: For readability, I tend to comment each primitive/unspecified value with its argument name:

```suggestion
      /* options */ undefined, /* host */ undefined, checkForErrors);
```

File: packages/compiler-cli/src/ngtsc/partial_evaluator/test/evaluator_spec.ts [hunk 6]
```diff
@@ -13,9 +13,8 @@ import {Reference} from '../../imports';
 import {DependencyTracker} from '../../incremental/api';
 import {FunctionDefinition, TsHelperFn, TypeScriptReflectionHost} from '../../reflection';
 import {getDeclaration, makeProgram} from '../../testing';
-import {DynamicValue} from '../src/dynamic';
+import {DynamicValue, EnumValue} from '../index';
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/partial_evaluator/test/evaluator_spec.ts]: Don't import from `index`, use a direct import to the file.

File: packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts [hunk 7]
```diff
@@ -8,29 +8,45 @@
...
 
+/**
+ * Resolves the specified builtin declaration to a resolved value. For example,
+ * the known JavaScript global `Object` will resolve to a `Map` that provides the
+ * `assign` method with a builtin function. This enables evaluation of `Object.assign`.
+ */
+export function resolveBuiltinDeclaration(builtinDecl: BuiltinDeclaration): ResolvedValue {
+  switch (builtinDecl) {
+    case BuiltinDeclaration.JsGlobalObject:
+      return new Map([['assign', new ObjectAssignBuiltinFn()]]);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts]: Rather than construct a new `Map` each time, let's have a shared global one.

**Review by alxhub — CHANGES_REQUESTED** [review]

File: packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts (hunk 2 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts]: Good idea. Worth mentioning: I took this from the `TsHelperFn` code 😄 I'll do the same there too.

File: packages/compiler-cli/src/ngtsc/partial_evaluator/test/utils.ts (hunk 5 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/partial_evaluator/test/utils.ts]: Sure, sounds good to me 👍

File: packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts (hunk 7 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts]: I had that initially, but figured that one could also argue that it unnecessarily allocates memory since it runs as a top-level module side-effect. Should be minimal though, so let's do it.

**Review by gkalpak — CHANGES_REQUESTED** [review]
Nice! This approach turned out quite clean :tada:

File: packages/compiler-cli/src/ngtsc/partial_evaluator/src/ts_helpers.ts [hunk 8]
```diff
@@ -10,17 +10,29 @@ import * as ts from 'typescript';
...
     case TsHelperFn.Spread:
     case TsHelperFn.SpreadArrays:
       return evaluateTsSpreadHelper(node, args);
+    case TsHelperFn.Assign:
+      // Use the same implementation we use for `Object.assign`. Semantically these
+      // functions are the same, so they can also share the same evaluation code.
+      return objectAssignBuiltinFn.evaluate(node, args);
     default:
-      throw new Error(`Cannot evaluate unknown helper ${helper} inline`);
+      throw new Error(`Cannot evaluate unknown helper ${TsHelperFn[helper]} inline`);
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/partial_evaluator/src/ts_helpers.ts]: I did this upon a review comment. I don't think it's necessarily to print out the name, so let's just keep it like before, and do the same for `BuiltinDeclaration`.

File: packages/compiler-cli/src/ngtsc/reflection/src/host.ts [hunk 9]
```diff
@@ -340,6 +340,20 @@ export enum TsHelperFn {
    * Indicates the `__spreadArrays` function.
    */
   SpreadArrays,
+  /**
+   * Indicates the `__assign` function.
+   */
+  Assign,
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/reflection/src/host.ts]: I'm using the chronological order 😄 Will change!

File: packages/compiler-cli/ngcc/src/host/esm2015_host.ts [hunk 10]
```diff
@@ -353,6 +353,25 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
       }
     }
 
+    // We detect if the resolved declaration refers to the JavaScript global `Object` by checking
+    // if the declaration node is named `Object`, resolves to a type `ObjectConstructor`, and has
+    // been defined in a default library source file.
+    if (superDeclaration && superDeclaration.node !== null &&
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/host/esm2015_host.ts]: In the compiler we tend to use explicit `null`/`undefined` checks, even if's more verbose.

```suggestion
    if (superDeclaration !== null && superDeclaration.node !== null &&
```

And yes, ngcc doesn't follow this convention very closely 😅

File: packages/compiler-cli/ngcc/src/host/esm2015_host.ts [hunk 11]
```diff
@@ -353,6 +353,25 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
...
+        ts.isVariableDeclaration(superDeclaration.node) &&
+        ts.isIdentifier(superDeclaration.node.name) &&
+        superDeclaration.node.name.text === 'Object' && superDeclaration.node.type &&
+        ts.isTypeReferenceNode(superDeclaration.node.type) &&
+        ts.isIdentifier(superDeclaration.node.type.typeName) &&
+        superDeclaration.node.type.typeName.text === 'ObjectConstructor' &&
+        this.src.program.isSourceFileDefaultLibrary(superDeclaration.node.getSourceFile())) {
```
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/ngcc/src/host/esm2015_host.ts]: I agree with George here, let's put this whole thing in a function so we can have some early returns and assign intermediate variables where it makes sense.

File: packages/compiler-cli/ngcc/src/host/esm5_host.ts [hunk 12]
```diff
@@ -8,10 +8,10 @@
 
 import * as ts from 'typescript';
 
-import {ClassDeclaration, ClassMember, ClassMemberKind, Declaration, Decorator, FunctionDefinition, Parameter, TsHelperFn, isNamedVariableDeclaration, reflectObjectLiteral} from '../../../src/ngtsc/reflection';
+import {ClassDeclaration, ClassMember, ClassMemberKind, Declaration, Decorator, FunctionDefinition, isNamedVariableDeclaration, Parameter, reflectObjectLiteral, TsHelperFn,} from '../../../src/ngtsc/reflection';
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/host/esm5_host.ts]:
```suggestion
import {ClassDeclaration, ClassMember, ClassMemberKind, Declaration, Decorator, FunctionDefinition, isNamedVariableDeclaration, Parameter, reflectObjectLiteral, TsHelperFn} from '../../../src/ngtsc/reflection';
```

File: packages/compiler-cli/src/ngtsc/partial_evaluator/test/utils.ts [hunk 13]
```diff
@@ -16,17 +16,20 @@ import {getDeclaration, makeProgram} from '../../testing';
 import {ForeignFunctionResolver, PartialEvaluator} from '../src/interface';
 import {ResolvedValue} from '../src/result';
 
-export function makeExpression(code: string, expr: string, supportingFiles: TestFile[] = []): {
+export function makeExpression(
+    code: string, expr: string, supportingFiles: TestFile[] = [], checkForErrors?: boolean): {
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/partial_evaluator/test/utils.ts]: Is this change still needed?

File: packages/compiler-cli/ngcc/src/host/esm2015_host.ts [hunk 14]
```diff
@@ -353,6 +353,25 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
...
+      return {
+        builtin: BuiltinDeclaration.JsGlobalObject,
+        expression: id,
+        viaModule: null,
+        node: null,
+      };
+    }
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/ngcc/src/host/esm2015_host.ts]: I think the idea is to keep ngtsc as lightweight as possible?

File: packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts [hunk 15]
```diff
@@ -40,3 +40,27 @@ export class ArrayConcatBuiltinFn extends BuiltinFn {
...
+    } else if (!(target instanceof Map)) {
+      return DynamicValue.fromUnsupportedSyntax(node);
+    }
+    for (let i = 1; i < args.length; i++) {
+      const source = args[i];
+      if (source instanceof DynamicValue) {
+        return DynamicValue.fromDynamicInput(node, source);
+      } else if (!(source instanceof Map)) {
+        return DynamicValue.fromUnsupportedSyntax(node);
+      }
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts]: great idea. thx!

File: packages/compiler-cli/ngcc/src/host/esm2015_host.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/ngcc/src/host/esm2015_host.ts]: The idea was that we just want to support `__assign` and `Object.assign` in ngcc for backwards compatibility. Previously the spread operator worked fine because the actual spread AST got put into the `metadata.json`. So, explicitly using `Object.assign` didn't work in ngc either.

I don't feel strongly about it. I just don't feel like there is a compelling reason to put it into ngtsc. Happy to change it if you all feel strongly about it.

File: packages/compiler-cli/ngcc/src/host/esm5_host.ts [hunk 16]
```diff
@@ -659,6 +659,8 @@ function getTsHelperFn(node: ts.NamedDeclaration): TsHelperFn|null {
       return TsHelperFn.Spread;
     case '__spreadArrays':
       return TsHelperFn.SpreadArrays;
+    case '__assign':
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/ngcc/src/host/esm5_host.ts]: 😄

File: packages/compiler-cli/ngcc/src/host/esm5_host.ts [hunk 17]
```diff
@@ -8,10 +8,10 @@
 
 import * as ts from 'typescript';
 
-import {ClassDeclaration, ClassMember, ClassMemberKind, Declaration, Decorator, FunctionDefinition, Parameter, TsHelperFn, isNamedVariableDeclaration, reflectObjectLiteral} from '../../../src/ngtsc/reflection';
+import {ClassDeclaration, ClassMember, ClassMemberKind, Declaration, Decorator, FunctionDefinition, isNamedVariableDeclaration, Parameter, reflectObjectLiteral, TsHelperFn,} from '../../../src/ngtsc/reflection';
 import {getNameText, hasNameIdentifier, stripDollarSuffix} from '../utils';
 
-import {Esm2015ReflectionHost, ParamInfo, getPropertyValueFromSymbol, isAssignment, isAssignmentStatement} from './esm2015_host';
+import {Esm2015ReflectionHost, getPropertyValueFromSymbol, isAssignment, isAssignmentStatement, ParamInfo} from './esm2015_host';
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/ngcc/src/host/esm5_host.ts]: wouldn't clang-format complain if it would be case-sensitive? I'm changing it, but I don't think it's a good rule if it's not enforced somehow. FWIW: Webstorm ordered it that way for me.

File: packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin_declaration.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+ * the known JavaScript global `Object` will resolve to a `Map` that provides the
+ * `assign` method with a builtin function. This enables evaluation of `Object.assign`.
+ */
+export function resolveBuiltinDeclaration(builtinDecl: BuiltinDeclaration): ResolvedValue {
+  switch (builtinDecl) {
+    case BuiltinDeclaration.JsGlobalObject:
+      return jsGlobalObjectValue;
+    default:
+      throw new Error(
+          `Cannot resolve unknown builtin declaration ${BuiltinDeclaration[builtinDecl]}`);
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin_declaration.ts]: See comment above.

File: packages/compiler-cli/src/ngtsc/partial_evaluator/test/utils.ts (hunk 13 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/partial_evaluator/test/utils.ts]: good catch. removed

File: packages/compiler-cli/src/ngtsc/reflection/src/host.ts [hunk 19]
```diff
@@ -340,6 +340,20 @@ export enum TsHelperFn {
...
+}
+
+/**
+ * Possible declarations which are known.
+ */
+export enum BuiltinDeclaration {
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/reflection/src/host.ts]: done. agreed

File: packages/compiler-cli/ngcc/src/packages/entry_point_bundle.ts [hunk 20]
```diff
@@ -50,8 +50,7 @@ export function makeEntryPointBundle(
   const rootDir = entryPoint.package;
   const options: ts.CompilerOptions = {
     allowJs: true,
-    maxNodeModuleJsDepth: Infinity,
-    noLib: true, rootDir, ...pathMappings
+    maxNodeModuleJsDepth: Infinity, rootDir, ...pathMappings
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/ngcc/src/packages/entry_point_bundle.ts]: I didn't. It has been enabled in the past too, but just seems to have been disabled when the mocked file system has been introduced. Can surely do some measurements if needed.

**Comment by devversion** [comment]
@gkalpak @JoostK Thanks for reviewing. Addressed the feedback.

File: packages/compiler-cli/ngcc/src/host/esm2015_host.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/compiler-cli/ngcc/src/host/esm2015_host.ts]: Makes sense. Let's keep it in ngcc only for now.

File: packages/compiler-cli/ngcc/src/host/esm5_host.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/compiler-cli/ngcc/src/host/esm5_host.ts]: I am pretty sure clang-format used to put uppercase imports before lowercase ones.
Maybe it doesn't care any more 🤷‍♂

File: packages/compiler-cli/ngcc/src/packages/entry_point_bundle.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/compiler-cli/ngcc/src/packages/entry_point_bundle.ts]: Fair enough. If it has been there before, it can't be too bad :grin:

@petebacondarwin, any idea why was this removed in https://github.com/angular/angular/commit/16d7dde2ad89eb4610c74da63e0e4b00afb8b577#diff-ac19368dcc489893e10fcaa6ff9436aeR48 (just to make sure there is nothing we should take into account)?

**Review by gkalpak — APPROVED** [review]
LGTM! Great work, @devversion :100:

(I still wonder about `noLib` (https://github.com/angular/angular/pull/34661#discussion_r367637113), but that shouldn't block the PR from landing.)

**Review by JoostK — APPROVED** [review]
Excellent work!

**Comment by devversion** [comment]
@gkalpak I can do some measurements, but I doubt that `noLib` affects performance significantly.

**Comment by gkalpak** [comment]
I am not concerned about performance. My question (https://github.com/angular/angular/pull/34661#discussion_r367637113) was more about whether thare might any other reason to add/remove it.

**Comment by devversion** [comment]
Fair enough. Maybe @petebacondarwin remembers why the change has been made. IIRC we already chatted about this on Slack, and it seemed like it was because of the mocked file system which did not contain the TypeScript default library files.

**Comment by petebacondarwin** [comment]
I believe I did add it for one of two reasons:
a) it would be faster to process if we loaded in as few files into the program as possible
b) the mock file systems did not include these files

If everything works as expected with this removed, and there is no significant performance penalty, then I am happy.

File: packages/compiler-cli/ngcc/src/packages/entry_point_bundle.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/ngcc/src/packages/entry_point_bundle.ts]: Probably because it was a fair bit of work to introduce the default library typings into the abstract filesystem.

I doubt including an extra .d.ts file or two will cause any performance issues.

File: packages/compiler-cli/src/ngtsc/partial_evaluator/src/known_declaration.ts [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+ * Resolves the specified known declaration to a resolved value. For example,
+ * the known JavaScript global `Object` will resolve to a `Map` that provides the
+ * `assign` method with a builtin function. This enables evaluation of `Object.assign`.
+ */
+export function resolveKnownDeclaration(decl: KnownDeclaration): ResolvedValue {
+  switch (decl) {
+    case KnownDeclaration.JsGlobalObject:
+      return jsGlobalObjectValue;
+    default:
+      throw new Error(`Cannot resolve known declaration. Received: ${decl}.`);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/partial_evaluator/src/known_declaration.ts]:
```suggestion
      throw new Error(`Cannot resolve known declaration. Received: ${KnownDeclaration[decl]}.`);
```
(prints the name, not the numeric value)

**Review by alxhub — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:291025335:BASE:291025371:1579730515280:5bc45ba7)

**Comment by AndrewKushnir** [comment]
@devversion, could you please rebase this PR? Thank you.

**Comment by devversion** [comment]
@AndrewKushnir Rebased!

**Review by gkalpak — APPROVED** [review]
